src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs

```diff
@@ -260,5 +261,239 @@ public static bool HasAllSet(this BitArray bitArray)
+                // concatenation of the integral and fractional parts are sequence equal.
+                // Under this scheme the number 0 is represented by a pair of empty spans.
+
+                Debug.Assert(span.Length > 0);
+
+                if (span[0] == '-')
+                {
+                    isNegative = true;
+                    span = span.Slice(1);
+                }
+                else
+                {
+                    Debug.Assert(char.IsDigit((char)span[0]), "leading plus not allowed in valid JSON numbers.");
+                    isNegative = false;
+                }
+
+#if NET
+                int i = span.IndexOfAny(s_decimalPointOrExponent);
+#else
+                int i = span.IndexOfAny((byte)'.', (byte)'e', (byte)'E');
+#endif
+                if (i < 0)
+                {
+                    integral = span;
+                    fractional = default;
+                    exponent = 0;
+                    goto Normalize;
+                }
+
+                integral = span.Slice(0, i);
+
+                if (span[i] == '.')
+                {
+                    span = span.Slice(i + 1);
+#if NET
+                    i = span.IndexOfAny(s_exponent);
+#else
+                    i = span.IndexOfAny((byte)'e', (byte)'E');
+#endif
+
+                    if (i < 0)
+                    {
+                        fractional = span;
+                        exponent = 0;
+                        goto Normalize;
+                    }
+
+                    fractional = span.Slice(0, i);
+                }
+                else
+                {
+                    fractional = default;
+                }
+
+                Debug.Assert(span[i] is (byte)'e' or (byte)'E');
+                bool success = Utf8Parser.TryParse(span.Slice(i + 1), out exponent, out _);
+                Debug.Assert(success);
+
+            Normalize:
+                if (integral[0] == '0')
+                {
+                    // Normalize "0" to the empty span.
+                    Debug.Assert(integral.Length == 1, "Leading zeros not permitted in JSON numbers.");
+                    integral = default;
+                }
+
+                if (IndexOfFirstTrailingZero(fractional) is >= 0 and int iz)
+                {
+                    // Trim all trailing zeros from the fractional part.
+                    fractional = fractional.Slice(0, iz);
+                }
+
+                if (fractional.IsEmpty && IndexOfFirstTrailingZero(integral) is >= 0 and int fz)
+                {
+                    // There is no fractional part, trim all trailing zeros from
+                    // the integral part and increase the exponent accordingly.
+                    exponent += integral.Length - fz;
+                    integral = integral.Slice(0, fz);
+                }
+
+                // Normalize the exponent by subtracting the length of the fractional part.
+                exponent -= fractional.Length;
```

Comment:
Some of these operations are likely to be more expensive than necessary due to all of these being by ref parameters rather than locals.